File: apps/orchestrator/src/nodes/agentd/control-auth.test.ts
import { describe, expect, it } from 'vitest';

import { controlMac, controlNonce, validControlNonce, verifyControlMac } from './control-auth.js';

describe('agentd v2 control authentication', () => {
  it('matches the shared Go/TypeScript vector', () => {
    expect(
      controlMac({
        credential: '0123456789abcdef0123456789abcdef',
        role: 'server',
        nodeId: 'node-1234',
        clientNonce: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA',
        serverNonce: 'BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB',
        daemonVersion: '0.3.0',
        capabilities: ['pty', 'resize', 'scrollback'],
      }),
    ).toBe('pSdsQrZCMrYUFj85DnoQGcbUl2Jjr7kiehtn60Voruc');
  });

  it('generates 256-bit nonces and compares MACs safely', () => {
    const nonce = controlNonce();
    expect(validControlNonce(nonce)).toBe(true);
    expect(validControlNonce('short')).toBe(false);
    expect(verifyControlMac('same', 'same')).toBe(true);
    expect(verifyControlMac('same', 'different')).toBe(false);
  });
});

File: agentd/internal/controlauth/controlauth_test.go
package controlauth

import "testing"

func TestSharedHandshakeVector(t *testing.T) {
	got := MAC(
		"0123456789abcdef0123456789abcdef",
		"server",
		"node-1234",
		"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
		"BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB",
		"0.3.0",
		[]string{"pty", "resize", "scrollback"},
	)
	const want = "pSdsQrZCMrYUFj85DnoQGcbUl2Jjr7kiehtn60Voruc"
	if got != want {
		t.Fatalf("shared vector mismatch: got %q want %q", got, want)
	}
}

func TestNonceShapeAndVerification(t *testing.T) {
	nonce, err := Nonce()
	if err != nil || !ValidNonce(nonce) {
		t.Fatalf("bad nonce %q: %v", nonce, err)
	}
	if ValidNonce("short") {
		t.Fatal("accepted short nonce")
	}
	if !Verify("same", "same") || Verify("same", "different") {
		t.Fatal("MAC comparison failed")
	}
}
